python/src/open_discourse/steps/factions/add_abbreviations_and_ids.py

```diff
@@ -1,82 +1,174 @@
-        "Fraktion der SPD (Gast)": "SPD",
-        "Fraktion der Sozialdemokratischen Partei Deutschlands": "SPD",
-        "Fraktionslos": "Fraktionslos",
-        "Gruppe Bündnis 90/Die Grünen": "Bündnis 90/Die Grünen",
-        "Gruppe BSW - Bündnis Sahra Wagenknecht - Vernunft und Gerechtigkeit": "BSW",
-        "Gruppe Deutsche Partei": "DP",
-        "Gruppe Die Linke": "DIE LINKE.",
-        "Gruppe Kraft/Oberländer": "KO",
-        "Gruppe der Partei des Demokratischen Sozialismus": "PDS",
-        "Gruppe der Partei des Demokratischen Sozialismus/Linke Liste": "PDS",
-        "Südschleswigscher Wählerverband": "SSW",
-        "Gast": "Gast",
-        "Gruppe Nationale Rechte": "NR",
-    }
-
-    factions.insert(0, "abbreviation", "")
-    factions["abbreviation"] = factions["faction_name"].apply(
-        lambda x: abbreviations_dict[x]
+    Args:
+        faction_name (str): The full faction name to look up in the FACTION_ABBREVIATIONS dictionary
+
+    Returns:
+        str: The standardized abbreviation if the faction name is found in FACTION_ABBREVIATIONS,
+             otherwise returns the original faction_name as a fallback
+    """
+    if faction_name in FACTION_ABBREVIATIONS:
+        return FACTION_ABBREVIATIONS[faction_name]
+    else:
+        # Return the faction name itself as a fallback
+        return faction_name
+
+
+def add_abbreviations_to_factions(factions_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Adds abbreviations to faction names based on predefined mapping.
+
+    Args:
+        factions_df (pd.DataFrame): DataFrame containing faction names in the 'faction_name' column
+
+    Returns:
+        pd.DataFrame: DataFrame with an additional 'abbreviation' column at index 0. The original
+                      data is preserved, and for faction names without a standard abbreviation,
+                      the original faction name is used as the abbreviation.
+
+    Raises:
+        KeyError: If 'faction_name' column is missing from the input DataFrame
+    """
+    logger.info("Adding abbreviations to factions")
+
+    # Create a copy to avoid modifying the input DataFrame
+    result_df = factions_df.copy()
+
+    # Insert new column at the beginning
+    result_df.insert(0, "abbreviation", "")
+
+    # Track missing factions
+    missing_factions = []
+
+    # Helper function to track missing factions while getting abbreviation
+    def track_and_get_abbreviation(faction_name):
+        abbreviation = _get_abbreviation(faction_name)
+        if abbreviation == faction_name and faction_name not in FACTION_ABBREVIATIONS:
+            missing_factions.append(faction_name)
+        return abbreviation
+
+    # Apply the function to each faction name
+    result_df["abbreviation"] = result_df["faction_name"].apply(
+        track_and_get_abbreviation
     )
 
-    unique_abbreviations = np.unique(factions["abbreviation"])
+    # Log any missing factions as warnings
+    if missing_factions:
```

Comment:
we don't have to do it as a second step outside, why not directly do it inside `track_and_get_abbreviation` when you actually test for missing abbreviations?